### PR TITLE
Full CNN architecture adaptation, and update of the DataLoader

### PR DESCRIPTION
**Pull request by shedprog**
In the following PR working example of CNN with a queue-to-queue multiprocessing dataloader is introduced:

Currently, updated:
- [x] 2017v2/Training_p6.py with some modifications cloned to 2018v1/Training_v0p1.py
- [x] Adapted NN config variables (moved initialization from command.py to DataLoadet.get_config())
- [x] Added queue-to-queue multiprocessing, when one instance of c++ DataLoader is considered as a single-core worker

Todo checklist:
- [x] Merge with changes in scaling and weighting
- [x] Implement dataloader stop on fixed number of the batches

Technical details and benchmarks:
(The tests were performed at the GPU model: NVIDIA GeForce GTX 1080 Ti computeCapability: 6.1, global device memory was limited to 10 Gb.)
- The runtime profiling of the step gives the results:
image
**Average Step Time (sum): 684.7 ms**
All Others Time (including I/O ops): 122.0 ms
Kernel Launch Time: 64.6 ms
Host Compute Time:7.9 ms
Device Compute Time: 490.1 ms

- At the moment, I do not observe significant increase of host memory during training, but to be checked on the bigger scale. Current plot is done for ~ 1000 batches, 500 taus each, 10 elements in queue,  5 dataloader workers
image

### CODE REVIEW

File: Training/python/DataLoader.py [hunk 1]
```diff
@@ -1,228 +1,259 @@
...
-            thread = Thread(target=LoaderThread,
-                     args=( file_entries, queue, self.net_config, self.batch_size, self.chunk_size,
-                            return_truth, return_weights, self.return_grid ))
-            thread.daemon = True
-            thread.start()
-            for step_id in range(steps_per_epoch):
-                item = queue.get()
-                yield item
-            thread.join()
+    def ListToVector(l, elem_type):
```
Comment (kandrosov) [inline Training/python/DataLoader.py:93]: `ListToVector` is a general purpose function, move its definition outside of DataLoader

File: Training/python/DataLoader.py [hunk 2]
```diff
@@ -1,228 +1,259 @@
...
+                mp.Process(target = LoaderThread, 
+                        args = (queue_out, queue_files, self.input_grids,
+                                self.batch_size, self.n_inner_cells, self.n_outer_cells,
+                                self.n_flat_features, self.n_grid_features, self.tau_types,
+                                return_truth, return_weights)))
+                processes[-1].deamon = True
+                processes[-1].start()
+
+
+            while True:
```
Comment (kandrosov) [inline Training/python/DataLoader.py:191]:
```python
while finish_counter < self.n_load_workers:
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Training/python/DataLoader.py [hunk 3]
```diff
@@ -1,228 +1,283 @@
...
+                     (batch_size, _n_cells, _n_cells, n_grid_features[fname])) for fname in group ],
+                    axis=-1
+                    ),dtype=tf.float32)
+                )
+        return _X
+
+    _dl_worker = R.DataLoader()
+    _req_file = True
+
+    while batch_counter.value < n_batches or n_batches == -1:
```
[the remark this answers is not in the document]
Comment (kandrosov) [reply Training/python/DataLoader.py:41]: it was a bug in the proposed `QueueEx.put` implementation. Here is the corrected version:
```python
    def put(self, item, retry_interval=0.3):
        while True:
            with self.n_puts.get_lock():
                if self.n_puts.value >= self.max_n_puts:
                    return False
                try:
                    self.mp_queue.put(item, False)
                    self.n_puts.value += 1
                    return True
                except FullException:
                    pass
            time.sleep(retry_interval)

```

File: Training/interface/DataLoader_main.h [hunk 4]
```diff
@@ -200,47 +200,63 @@ class DataLoader {
...
-        hasData = true;
+        fullData = true;
         return true;
     }
 
-    size_t GetEntries () { return end_entry - current_entry; }
-
-    std::shared_ptr<Data> LoadData() {
-      if(!hasData)
+    Data LoadData() {
```
[the remark this answers is not in the document]
Comment (shedprog) [reply Training/interface/DataLoader_main.h:252]: I found the following options:
```c++
std::unique_ptr<Data> LoadData() {
...
return std::move(data);
}
``` 
or
```c++
Data* LoadData() {
...
return std::move(data.get());
}

**Comment by shedprog** [comment]
After final modification, few more benchmarks were done:

Training is run with 13GB GPU available on NVIDIA Tesla P100-PCIE-16GB
with `n_batches=1000, n_batches_val=200, max_queue_size=10, n_load_workers=4` epochwise the training looks stable (looking at general loss function)
image
the RAM consumption in the order of 12GB so, if it is run within htcondor job I would require ~15 GB

The profiling shows the following results (they are worst comparing to what was measured before, there are some peaks during which for some reasons dataloader does not response):
Average Step Time: 862.1 ms
All Others Time: 240.4 ms
Input Time: 148.9 ms
Kernel Launch Time: 59.9 ms
Device Compute Time: 400.3 ms
image

**Comment by shedprog** [comment]
I also implemented an alternative python dataloader-wrapper implementation https://github.com/shedprog/TauMLTools_newloader/blob/master/python/DataLoaderInter.py which is based on  tf.dataset.interleave functional:
https://github.com/shedprog/TauMLTools_newloader/blob/c6e44af77165806ddbfe6c58717c833584445385/python/DataLoaderInter.py#L231-L249 

So far, this implementation does not outperform mp.Queue-based that we have under the current master branch.

**Review by kandrosov — APPROVED** [review]

**Comment by kandrosov** [comment]
I propose that we keep tuning of `tf.dataset.interleave` based implementation for future iterations. I'm going to merge this PR. Thank you!
